### ISSUE TITLE
Interoperating with C arrays (and C array fields)

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#10162]
As a Chapel programmer, I'd like a way to interoperate with arrays declared within my C code, particularly as array fields, in-place (i.e., no copies into Chapel).  Ideally, I'd like to access these arrays as though they were Chapel arrays, supporting things like forall loops, function promotion, slicing, and such.  Here are four cases for consideration.  How would I express these as `extern records` in Chapel and what would/could the implementation need to do to support treating them as normal Chapel arrays within Chapel?

```c
struct {
  int x[100];
};
```

```c
#define ASIZE 100
struct {
  int x[ASIZE];
};
```

```c
struct {
  int a;
  int b;
  int c[]; // Variable sized.
};
```

```c
struct {
  int *x; // Pointer to buffer of consecutive integers
};
```

**Comment by bradcray** [comment]
I mentioned this in a meeting a few weeks back, making a loose connection to Lydia's support for an external array domain map in the Python/Cython interoperability context.

These cases are complicated somewhat by being within a record since it suggests we can't wrap arbitrary `_array` record wrappers around them (at least in-place).  In one brainstorming session I wondered whether we'd need a hash table from external C pointers to `_array` descriptors (or domain map classes) that supported Chapel functionality on them.

**Comment by bradcray** [comment]
Tagging @lydia-duncan and @daviditen on this issue because things you're exploring arguably nudge up against it.

**Comment by lydia-duncan** [comment]
Seen, but not sure I have a good answer to the question yet

**Comment by mppf** [comment]
Here are the answers as I see them after PR #12156:

### case one
```c
typedef struct {
  int x[100];
} case_one;
```

``` chapel
extern record case_one {
  var x:c_array(c_int, 100);
}
```

### case two
```c
#define ASIZE 100
typedef struct {
  int x[ASIZE];
} case_two;
```

``` chapel
param ASIZE=100;
// It would be nice to have `extern param` for such #defines 
// but we don't yet and that's not natural/easy to implement with
// the C backend (but it is natural/easy to implement with
// the LLVM backend).
extern record case_two {
  var x:c_array(c_int, ASIZE);
}
```

### case three
```c
typedef struct {
  int a;
  int b;
  int c[]; // Variable sized.
} case_three;
```

``` chapel
extern record case_three {
  // We could make c_array(0) disable bounds checking
  // and work for variable sized/zero length C structure
  // members, but that's not implemented yet.
  var x:c_array(c_int, 0);
}
```

### case four
```c
typedef struct {
  int *x; // Pointer to buffer of consecutive integers
} case_four;
```

``` chapel
extern record case_four {
  var x:c_ptr(c_int);
}
```

**Comment by mppf** [comment]
@bradcray - I wonder if we should close this issue now?

I can see two follow-on efforts, if we decide to undertake them:
 * support varargs structure members (e.g. with `c_array(T, 0)`)
 * allow --llvm to turn #defines into `extern param`

But if these are even issue-worthy they probably deserve their own issues. Thoughts?

**Comment by bradcray** [comment]
I think I pinged the original user to see if they were satisfied but didn't hear back.  Let me try again.

**Comment by bradcray** [comment]
Tagging @stonea on this issue, which was one of the original motivators for `c_array` and because it mentions "as though they were Chapel arrays" suggesting that providing productivity features above and beyond C's (as suggested in #17946) could be attractive.  That's not to say "we must do it", simply a data point to consider.

Also tagging @mppf who I believe asked "Why do we have these?" the other week when reviewing the CPtr module (though maybe rhetorically because then he answered the question himself).
